### DISCUSSION TITLE
conform to a protocol with descriptors

### DISCUSSION THREAD

**Discussion by saryou** [microsoft/pyright#3994] [General]
I want to make a class conform to a protocol both the class object and instances. And I tried to achieve this by using descriptor like below but it didn't work.

```
from typing import Protocol, Callable, Any


class AProtocol(Protocol):
    @property
    def a(self) -> Callable[[], int]:
        ...


def take_a(a: AProtocol) -> int:
    return a.a()


class DesctiptorA:
    def __get__(self, instance: Any, owner: Any = None) -> Callable[[], int]:
        return lambda: 1


class A:
    a = DesctiptorA()


print(take_a(A))
print(take_a(A()))
```

I think it should be ok. Or is there any other way to make it work?

**Comment by erictraut** [comment]
Your approach won't work because `AProtocol` specifies that a compatible object or class must support a `property` called `a`. A `property` is not the same as a descriptor. They are related, in that a `property` is implemented using a descriptor protocol, but `property` provides a bunch of additional functionality and behavior. 

This isn't a bug, so I'm going to convert this to a discussion topic.

**Reply by saryou** [reply]
Thanks for your reply!

I found that mypy has open issue about protocols and descriptors:
https://github.com/python/mypy/issues/5481
